### PR TITLE
xtimer: fix mutex unlocking in _mutex_timeout

### PR DESCRIPTION
**Pull request by vincent-d**
Changes proposed in #6427.

### CODE REVIEW

File: sys/xtimer/xtimer.c [hunk 1]
```diff
@@ -233,8 +233,11 @@ static void _mutex_timeout(void *arg)
     mutex_thread_t *mt = (mutex_thread_t *)arg;
 
     mt->timeout = 1;
-    sched_set_status(mt->thread, STATUS_PENDING);
     list_remove(&mt->mutex->queue, (list_node_t *)&mt->thread->rq_entry);
+    if (mt->mutex->queue.next == NULL) {
```
Comment (lebrush) [inline sys/xtimer/xtimer.c]: Really good catch! :+1: 

I ignored that the `MUTEX_LOCK` is not always the tip of the list! It is only when the mutex is owned by one thread. If there are more threads waiting, the tip is NULL. I guess this is done to be able to use `thread_add_to_list` cleanly.

There's one race condition missing. Otherwise if the timer fires between setting the timer and locking the mutex might create inconsistencies in the mutex queue. Just test that `list_remove` really removed a node

```C
list_node_t* node = list_remove(...)
if (node != NULL && mt->mutex->queue.next == NULL) {
    // we were the last node in the queue but the mutex shall still be owned by someone else
    mt->mutex->queue.next = MUTEX_LOCKED;
}
```

File: sys/xtimer/xtimer.c [hunk 2]
```diff
@@ -233,8 +233,12 @@ static void _mutex_timeout(void *arg)
     mutex_thread_t *mt = (mutex_thread_t *)arg;
 
     mt->timeout = 1;
+    list_node_t *node = list_remove(&mt->mutex->queue,
+                                    (list_node_t *)&mt->thread->rq_entry);
+    if (node != NULL && mt->mutex->queue.next == NULL) {
```
Comment (OlegHahm) [inline sys/xtimer/xtimer.c]: Please but each condition into parentheses.

Comment (vincent-d) [inline sys/xtimer/xtimer.c]: addressed

**Comment by vincent-d** [comment]
@OlegHahm comment addressed and squashed

**Comment by lebrush** [comment]
Please wait before merging this one

**Comment by OlegHahm** [comment]
@lebrush, waiting for what?

**Comment by lebrush** [comment]
Sorry for the delay. 

After reviewing the code once more I made a table with the following corner cases: what happens if the timer fires before calling `mutex_lock` and what if fires after calling `mutex_lock`. Since the code within `xtimer_mutex_lock_timeout` is not atomic we have some possible race conditions.

| Mutex was | Before `mutex_lock` | After `mutex_lock` | 
|---|---|---|
| UNLOCKED | mutex: `LOCKED (1)`, return: -1 | - | 
| LOCKED (1) | `mutex_lock` blocks until unlocked; return -1 | mutex: `LOCKED (1)`, return -1 (with this patch) |
| LOCKED (2+) | `mutex_lock` blocks until unlocked; return -1 | mutex: `LOCKED (2+)`; return -1 |

This patch solves the case `LOCKED (1) | After mutex_lock`.

| Mutex was | Before `mutex_lock` | After `mutex_lock` | 
|---|---|---|
| UNLOCKED | WRONG | - | 
| LOCKED (1) | WRONG | OK |
| LOCKED (2+) | WRONG | OK |

The implementation of xtimer might decide to spin if the timeout is to small. Hence we need to solve the `Before` column as well. One solution would be to replace the `if (timeout != 0)` with:

```C
if (timeout > XTIMER_BACKOFF) 
```

With these two fixes then we are:

| Mutex was | Before `mutex_lock` | After `mutex_lock` | 
|---|---|---|
| UNLOCKED | OK | - | 
| LOCKED (1) | OK | OK |
| LOCKED (2+) | OK | OK |

The last consideration I could think of is the case that the timer triggers between the return of `mutex_lock` (isr safe) and `xtimer_remove` (isr safe). The mutex would have the correct state but the method would return -1. How probable is that this happens?

**Comment by lebrush** [comment]
TL;DR I would merge this one and contemplate the other cases in i.e. #6441

**Comment by OlegHahm** [comment]
What was the initial reason not to call `mutex_unlock()` in the callback?

**Comment by lebrush** [comment]
Thread A has the mutex, Thread B waits for it (highest prio). Thread C (lowest prio) sets a mutex lock with timeout. If the callback occurs before A has released the mutex and you call `mutex_unlock` thread A and B "have" the mutex now, right?

**Comment by OlegHahm** [comment]
Ah, I think I misunderstood `xtimer_mutex_lock_timeout()` (and I think the documentation is not very accurate). I thought the function would lock the timeout for a certain time (which doesn't have much use cases, I guess).

**Comment by lebrush** [comment]
Good point. Will improve the docs in #6441

File: sys/xtimer/xtimer.c [hunk 3]
```diff
@@ -233,8 +233,12 @@ static void _mutex_timeout(void *arg)
     mutex_thread_t *mt = (mutex_thread_t *)arg;
 
     mt->timeout = 1;
+    list_node_t *node = list_remove(&mt->mutex->queue,
+                                    (list_node_t *)&mt->thread->rq_entry);
+    if ((node != NULL) && (mt->mutex->queue.next == NULL)) {
+        mt->mutex->queue.next = MUTEX_LOCKED;
```
Comment (OlegHahm) [inline sys/xtimer/xtimer.c]: This has to be `MUTEX_LOCKED` because we assume that someone still owns the mutex, right?

Comment (lebrush) [inline sys/xtimer/xtimer.c]: Exactly.

When the mutex is unlocked, `queue.next` is NULL. ` mutex -> NULL`
When the mutex is locked by 1 thread `queue.next` is `MUTEX_LOCKED`.  `mutex -> LOCKED`
When the mutex is locked by 2+ threads `queue.next` is the `mt->thread->rq_entry` of a thread and the last `.next` is NULL. `mutex -> thread -> (thread -> ...) NULL`

If `node != NULL` we were removed from the list, hence 2+ threads where in the list. At the same time if `queue.next` is NULL means that we were the last thread in the list and we need to set `queue.next` to `MUTEX_LOCKED` as there's now 1 thread still owning the mutex.

Comment (vincent-d) [inline sys/xtimer/xtimer.c]: Yes, otherwise you could have locked it and then the timeout never fire.
